art/attacks/evasion/auto_projected_gradient_descent.py

```diff
@@ -108,112 +112,49 @@ def __init__(
-                    labels_ph=estimator._labels_ph,
-                    train=estimator._train,
-                    loss=self._loss_object,
-                    learning=estimator._learning,
-                    sess=estimator._sess,
-                    channels_first=estimator.channels_first,
-                    clip_values=estimator.clip_values,
-                    preprocessing_defences=estimator.preprocessing_defences,
-                    postprocessing_defences=estimator.postprocessing_defences,
-                    preprocessing=estimator.preprocessing,
-                    feed_dict=estimator._feed_dict,
-                )
+                        def __init__(self, from_logits, reduction="mean"):
+                            self.ce_loss = tf.keras.losses.CategoricalCrossentropy(
+                                from_logits=from_logits,
+                                reduction=tf.keras.losses.Reduction.NONE,
+                            )
+                            self.reduction = reduction
 
-            elif isinstance(estimator, TensorFlowV2Classifier):
-                import tensorflow as tf
+                        def __call__(self, y_true, y_pred):
+                            if self.reduction == "mean":
+                                return tf.reduce_mean(self.ce_loss(y_true, y_pred))
+                            if self.reduction == "sum":
+                                return tf.reduce_sum(self.ce_loss(y_true, y_pred))
+                            if self.reduction == "none":
+                                return self.ce_loss(y_true, y_pred)
+                            raise NotImplementedError()
 
-                if loss_type == "cross_entropy":
                     if is_probability(estimator.predict(x=np.ones(shape=(1, *estimator.input_shape)))):
-                        self._loss_object = tf.keras.losses.CategoricalCrossentropy(from_logits=False)
+                        _loss_object_tf: TensorFlowV2Loss = CrossEntropyLossV2(from_logits=False)
                     else:
-                        self._loss_object = tf.keras.losses.CategoricalCrossentropy(from_logits=True)
+                        _loss_object_tf = CrossEntropyLossV2(from_logits=True)
                 elif loss_type == "difference_logits_ratio":
                     if is_probability(estimator.predict(x=np.ones(shape=(1, *estimator.input_shape)))):
                         raise ValueError(  # pragma: no cover
                             "The provided estimator seems to predict probabilities. "
                             "If loss_type='difference_logits_ratio' the estimator has to to predict logits."
                         )
 
-                    class DifferenceLogitsRatioTensorFlowV2:
+                    class DifferenceLogitsRatioTensorFlowV2(TensorFlowV2Loss):
                         """
                         Callable class for Difference Logits Ratio loss in TensorFlow v2.
                         """
```

Comment:
Please change line 165 from `def __call__(self, y_true, y_pred):` to `def __call__(self, y_true: tf.Tensor, y_pred: tf.Tensor, *args, **kwargs) -> tf.Tensor:`